### ISSUE TITLE
self signed ssl certs not supported

### ISSUE THREAD

**Issue by phillip-white-sociomantic** [zulip/zulip-desktop#6]
I have a Zulip server that uses a self signed ssl cert. I am aware from this post https://groups.google.com/forum/#!topic/zulip-devel/e0PlJrePhYw that self signed certs are not supported. I am using the Windows desktop app from https://zulip.com/dist/apps/win/zulip-latest.exe

Can self signed ssl certs be supported on the native desktop app?

**Comment by lfranchi** [comment]
This has 2 parts: making it work on linux/windows (QNetworkAccessManager) and OS X (NSUrlRequest).

**Comment by shane-kearns** [comment]
The sslErrors signal is emitted when certificate verification fails:
http://doc.qt.io/qt-5/qnetworkaccessmanager.html#sslErrors

Whereas QSslConfiguration is the place to add additional CA certificates, for example to trust a company internal CA only for this application.

**Comment by paulolc** [comment]
Maybe a configurable directory where to place those internal CA certificates? Accepting self-signed certs is out of question?

**Comment by timabbott** [comment]
I think a configurable directory to place internal CA certs would be a pretty reasonable solution.

**Comment by lfranchi** [comment]
I suspect a configurable directory to place CA certs could be more work than adding a --insecure flag that makes the client ignore ssl cert verification errors.

Less secure as well though :)

**Comment by timabbott** [comment]
Yeah I think at least for some users, the secure solution is the only option that solves their problem, so I'd rather invest time into that...

**Comment by nicholasdavidson** [comment]
I've been working on this on and off in my spare time because my employer is investigating using zulip, my current plan of action is to add both a dialogue to allow the user to accept a self-signed certs and a place where certs you've accepted are stored. This should neatly solve both requirements as a ca-cert could be copied into the certs directory directly.

**Comment by timabbott** [comment]
Cool!  That seems pretty reasonable.

**Comment by lfranchi** [comment]
Great! Happy to review when ready :)

**Comment by slurdge** [comment]
@nicholasdavidson 
hi, I have a patch with an option to have insecure ssl with command line (i use it with my own server), i put it on hold I guess?

**Comment by nicholasdavidson** [comment]
I am still working on this but I've not had much spare time to finish it off and I don't want to block an improvement that's ready to go.

**Comment by timabbott** [comment]
@nicholasdavidson If you're not likely to find time to finish it soon, you should feel free to just post your partial branch so other folks can build off of it if they're excited to.

**Comment by lfranchi** [comment]
@slurdge we can include your insecure command-line switch as well as support the proper custom-ca-certs folder feature. So feel free to open a pull request :)

**Comment by lfranchi** [comment]
As of issue #12, now `--allow-insecure` works on windows and linux. OS X support still missing.

**Comment by slurdge** [comment]
@lfranchi Any idea of what is the missing bit for OSX? I don't have it myself, but maybe the patch is simple enough?

**Comment by WolkoDav** [comment]
Can anybody explain, we build project with latest changes and run with --allow-insecure arg. Unfortunately, we have white screen.

**Comment by slurdge** [comment]
@WolkoDav that's exactly how it should work. Do you have something in the logs ? What platform are you running on? I tested on both Linux and Windows and it worked with a self-signed certificate.

**Comment by WolkoDav** [comment]
Platform is Windows 10. 
Logs:

    QNetworkReplyImpl::_q_startOperation was called more than once
    qt.network.ssl: QSslSocket: cannot call unresolved function SSLv23_client_method
    qt.network.ssl: QSslSocket: cannot call unresolved function SSL_CTX_new
    qt.network.ssl: QSslSocket: cannot call unresolved function SSL_library_init
    qt.network.ssl: QSslSocket: cannot call unresolved function ERR_get_error
    qt.network.ssl: QSslSocket: cannot call unresolved function ERR_get_error

**Comment by phillip-white-sociomantic** [comment]
@timabbott Can you possibly include this fix in your ppa version? The build is not working for me, and i would rather have a deb for it anyway.

**Comment by timabbott** [comment]
@phillip-white-sociomantic I just sent a build of 0.5.1 to the Linux PPA.

**Comment by kpervoelzFHG** [comment]
@WolkoDav Did you solve your white screen problem? We are facing the same problem here with a self signed certificate.

The Mac app works just fine with our server but the latest windows app (zulip-latest.exe from today) just shows a white screen after startup. We are using Windows 7.

**Comment by slurdge** [comment]
@kpervoelzFHG I added a command line switch. Did you try to launch it with a --allow-insecure flag? Just running it as is will not work (and exhibit white screen).

**Comment by kpervoelzFHG** [comment]
yes, I did. I tried "zulip.exe --allow-insecure" from the command line to start zulip but got 

Error: "The certificate is self-signed, and untrusted" "The certificate is self-signed, and untrusted"...

in the log file. Am I using the command line option the right way?

**Comment by WolkoDav** [comment]
@kpervoelzFHG No, i reinstalled openssl and running client with --allow-insecure. Maybe i'm doing somethink wrong. i'm trying to build this with qt 5 and qt 4 after.

**Comment by slurdge** [comment]
@WolkoDav good remark. I build with qt5. Maybe the 0.5.1 build is with qt4, which has different semantics for ssl.

**Comment by WolkoDav** [comment]
@slurdge May be need some specific libraris for that? I'm trying use openssl libs for that, but its still not working...

**Comment by slurdge** [comment]
@WolkoDav frankly I don't have a clue. For Linux I just build with default libraries fro qt5.

**Comment by casimirextreme** [comment]
@all: Why not using a free ssl (only class 1 but still) certificate from https://www.startssl.com/ ?

zulip-desktop works without any hicups. The only thing to add is the intermediate certificate on the nginx config.

**Comment by kpervoelzFHG** [comment]
@casimirextreme Does this work also for internal server (no public domain name). As far as I understand the infos from startssl.com the class 1 certificates can be verified by domain and/or eMail. Do you know if I can use a class 1 certificate which is only validated by my eMail address for my internal zulip server without any problems?

**Comment by SewmiGitHub** [comment]
We resolved the issue (white screen after startup) by generating a certificate using our Internal CA to which all the domain computers are connected to. Then deleted the certificate and the key in the zulip server and replaced them with the ones generated by our Internal CA and renamed them to zulip.combined-chain.crt and zulip.key. Gave a 'service nginx restart'. Then zulip desktop worked just fine.. Hope this helps.

**Comment by Fhajad** [comment]
Should the --allow-insecure work on the current Windows build? We're currently just testing Zulip and don't have our own Internal CA.

**Comment by kpervoelzFHG** [comment]
As far as I can see, the option is supposed to work on the latest windows build (pull request was merged on Oct. 15th and latest windows build is from Oct. 16th) but we couldn't get it running here with a self signed certificate. On windows we are facing the "white screen problem".

**Comment by smewp** [comment]
I'm using a comodo positive SSL and have main server, mac client, ios clients running fine with no problem.  However, can't get the windows client beyond white screen, tested on several different windows 10 computers. Installed, reinstalled, several times, no help.  Is there a windows client troubleshooting guide?  Thanks all.

**Comment by ahhhkah** [comment]
We got the white screen problem as well using client version 0.5.1 and a self signed certificate (flagged with --allow-insecure on Windows 8.1).

**Comment by zaafar** [comment]
yes, same while screen problem in windows client (0.5.1) and self signed certificate.

**Comment by smewp** [comment]
To follow up, in our case I was not trying to use self signed certificate, but still had the white screen issue.  

Predictably it turns out I had misconfigured the ssl. Windows client v0.5.1 is working now after combining the .crt and .ca-bundle per instructions at below link and then installing it as /etc/ssl/certs/zulip.combined-chain.crt.

https://support.comodo.com/index.php?/Default/Knowledgebase/Article/View/789/37/certificate-installation-nginx

**Comment by zaafar** [comment]
I am able to run windows client (0.5.1) using self signed certificate by installing it into the internet explorer. This method is insecure but it works. Following are the steps to add the certificate into internet explorer.
- Run IE in administrator mode
- open your zulip server website on IE and click “Certificate Error” at the right of the address bar
- Click on “Install Certificate...”, then in the wizard, click “Next”
- On the next page select “Place all certificates in the following store”
- Click “Browse”, select “Trusted Root Certification Authorities”, and click “OK”
- Back in the wizard, click “Next”, then “Finish”
- now restart zulip desktop, hopefully it will work now.

**Comment by ahhhkah** [comment]
Thank you @zaafar. 

I managed to install the certificate via certmgr.msc into  “Trusted Root Certification Authorities” and now I am able to use the desktop client.

**Comment by Epighue** [comment]
@smewp Tried to follow the comodo article you recommended. Eg code was Example Syntax: cat www_yourdomain_com.crt www_yourdomain_com.ca-bundle > ssl-bundle.crt
Where did you get the .ca-bundle to combine for self signed cert? Thanks. New to this.

**Comment by Fhajad** [comment]
@Epighue @smewp wasn't using a self signed cert, only thrusted certs have a bundle for their certificate and them intermediate certificates. All the steps for creating a cert automatically are mentioned in the README.prod.md in the first section about "Installing Zulip in production". Just use those lines and it'll be taken care of automatically.

https://github.com/zulip/zulip/blob/master/README.prod.md

**Comment by rachkoud** [comment]
Hi, I know that is not secured, but for a test environment, you can trust the self-signed certificate, I was able to run zulip desktop on OS X and Windows with a self-signed certificate.

For OS X : [You can follow these steps](https://www.accuweaver.com/2014/09/19/make-chrome-accept-a-self-signed-certificate-on-osx/)

Fow Windows : [You can follow these steps](http://blogs.technet.com/b/sbs/archive/2007/04/10/installing-a-self-signed-certificate-as-a-trusted-root-ca-in-windows-vista.aspx)

Don't forget, when you generate the certificate (openssl req -new -key zulip.key -out server.csr) the common name  must be the hostname (localhost, hostname, or you can use a temporary domain zulip.domain.com and place it for test purpose in /etc/hosts for OS X or C:\Windows\System32\Drivers\etc\hosts for windows)

If your certificate is trusted and there is still an error with the certificate, it's certainly a [certificate name missmatch error](https://www.sslshopper.com/ssl-certificate-name-mismatch-error.html).

Hope it helps :smile:
